urld44fdf4fee_nadilsons_reflection_dsl_tgz-pJ8-br_com_bit_ideias_reflection_SimplePerformanceTestJ8/src/br/com/bit/ideias/reflection/scanner/PackageScanner.java

```diff
@@ -0,0 +1,729 @@
+            int foundIdx = -1;
+
+            strLoop:
+            for (int i = 0; i <= strLength - patLength; i++) {
+                for (int j = 0; j < patLength; j++) {
+                    String subPat = (String) pattDirs[pattIdxStart + j + 1];
+                    String subStr = (String) pathDirs[pathIdxStart + i + j];
+                    
+                    if (!matchStrings(subPat, subStr)) continue strLoop;
+                }
+                
+                foundIdx = pathIdxStart + i;
+                break;
+            }
+
+            if (foundIdx == -1) return false;
+
+            pattIdxStart = patIdxTmp;
+            pathIdxStart = foundIdx + patLength;
+        }
+
+        for (int i = pattIdxStart; i <= pattIdxEnd; i++) {
+            if (!pattDirs[i].equals("**")) return false;
+        }
+        
+        return true;
+    }
+
+    /**
+     * Find all resources in jar files that match the given location pattern
+     * via the Ant-style PathMatcher.
+     * @param rootDirResource the root directory as Resource
+     * @param subPattern the sub pattern to match (below the root directory)
+     * @return the Set of matching Resource instances
+     * @throws IOException in case of I/O errors
+     * @see java.net.JarURLConnection
+     * @see org.springframework.util.PathMatcher
+     */
+    protected Set<Resource> findPathMatchingJarResources(Resource rootDirResource, String subPattern) throws IOException {
+        URLConnection con = rootDirResource.getURL().openConnection();
+        JarFile jarFile = null;
+        String jarFileUrl = null;
+        String rootEntryPath = null;
+        boolean newJarFile = false;
+
+        if (con instanceof JarURLConnection) {
+            // Should usually be the case for traditional JAR files.
+            JarURLConnection jarCon = (JarURLConnection) con;
+            jarCon.setUseCaches(false);
+            jarFile = jarCon.getJarFile();
+            jarFileUrl = jarCon.getJarFileURL().toExternalForm();
+            JarEntry jarEntry = jarCon.getJarEntry();
+            rootEntryPath = (jarEntry != null ? jarEntry.getName() : "");
+        }
+        else {
+            // No JarURLConnection -> need to resort to URL file parsing.
+            // We'll assume URLs of the format "jar:path!/entry", with the protocol
+            // being arbitrary as long as following the entry format.
+            // We'll also handle paths with and without leading "file:" prefix.
+            String urlFile = rootDirResource.getURL().getFile();
+            int separatorIndex = urlFile.indexOf(getJarUrlSeparator());
+            if (separatorIndex != -1) {
+                jarFileUrl = urlFile.substring(0, separatorIndex);
+                rootEntryPath = urlFile.substring(separatorIndex + getJarUrlSeparator().length());
+                jarFile = getJarFile(jarFileUrl);
+            }
+            else {
+                jarFile = new JarFile(urlFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *jarFile*. The resource is referenced at lines 418, 428. The resource is closed at line 444. There are other execution paths that do not contain closure statements, for example, when an else condition at line 443 evaluates to true. Ensure *jarFile* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).